### ISSUE TITLE
Add a new :company-insert-function to capf to support lsp completion better

### ISSUE THREAD

**Issue by dgutov** [company-mode/company-mode#1423]
### Discussed in https://github.com/company-mode/company-mode/discussions/1422

<div type='discussions-op-text'>

<sup>Originally posted by **zbelial** October 28, 2023</sup>
I wish to improve user experience when using company-capf with lsp clients.  Here is my proposal.  I'll describe briefly how company + lsp capf works  and then explain how `:complete-function` can help here.

When you press ENTER after selecting a candidate from the popup, company + lsp clients' capf work in this way:
1. company deletes the prefix (if any) input already
2. company inserts candidate's content
3. company will call lsp capf's `:exit-function`, in which it deletes what has been inserted in step 2, and then inserts what's really desired (may need to expand a snippet via yasnippet)

here is an example. 
![图片](https://github.com/company-mode/company-mode/assets/6436456/e27c68af-ef54-4547-be85-45f9f95a5579)

I have input "add", this is the prefix.  Now I'll simulate step by step how company + lsp capf works
1. company deletes "add"
2. company inserts "addRound(BigDecimal a, BigDecimal b, Integer scale) : BigDecimal"
3. company deletes "addRound(BigDecimal a, BigDecimal b, Integer scale) : BigDecimal" and then inserts "addRound(a, b, scale)", 
Now you'll get what you really want (a snippet pre-filled with what jdt.ls thinks suitable as arguments)

![图片](https://github.com/company-mode/company-mode/assets/6436456/6cc8e4ea-152e-4a01-bfb7-0bc7ae0ae1d6)

Now I'll describe how `:complete-function` can improve this situation. With `:complete-function`, company + lsp capf works in this way:
1. company calls `:complete-function`, in which it deletes the prefix, and then inserts what's desired. (Very similar to what `:exit-function` does in this lsp case)

Because of no need to insert "addRound(BigDecimal a, BigDecimal b, Integer scale) : BigDecimal" and then delete it, and then insert what needed, compeltion will be more smooth and with less filicker. 

What do you think? </div>

**Comment by dgutov** [comment]
Previously mentioned in https://github.com/emacs-lsp/lsp-mode/pull/2711#issuecomment-798964844.

I do believe it's a good idea, but it probably won't improve anything in the way of flickering (insertion shouldn't flicker anyway), or completion performance, but it should make some code more straightforward.

**Comment by dgutov** [comment]
Basically, this either depends on or will be a part of #340.

And #426 will probably be implemented using this.

**Comment by zbelial** [comment]
Hi @dgutov , thanks for reply. 
When I said "flicker", I meant you can notice cursor jumping back and forth during the completion ( Sorry that I didn't find a more suitable word to describe the situation.)

I just recorded two gifs to demonstrate it, the first one **without** `:complete-function` enabled (In my company-mode fork, I added this `:complete-function`, and I used my own lsp client `lspce` when recording this two gifs), and in the second one, it's enabled. 

![flicker](https://github.com/company-mode/company-mode/assets/6436456/26a33de9-a2f9-46e7-b94c-cb252e225d8f)
  
![less fliker](https://github.com/company-mode/company-mode/assets/6436456/5ac98b67-9b96-4c43-b90a-68ecdebe4d7e)

So, will supporting `:company-insert-function` improve compeltion performance? Only a little IMO.  But I think it really improve user experience.

**Comment by minad** [comment]
FWIW I don't think such a `:complete-function` is an improvement over `:exit-function`, which justifies the addition. It is just another formulation of the `:exit-function`. The difference would be that the selected candidate is passed as argument.

There won't be any change regarding to flicker or performance, since the prefix deletion happen anyway. The only steps saved are the insertion and deletion, which should not lead to redisplay. There is a bug somewhere which leads to the flicker in the gif above, for example an undesired call to redisplay.

I had made similar arguments already in https://github.com/minad/corfu/issues/292.

**Comment by minad** [comment]
@zbelial Did you find out what exactly causes the redisplay? From your issues, it seems to be present in both Company and Corfu. Maybe an `inhibit-redisplay` around the deletion insertion call sequence would help?

**Comment by zbelial** [comment]
Hi @minad , I think it's an improvement given that it saves an insertation and a deletion, and the content of this insertation and deletion is longer ( sometimes much longer)  than the prefix.  And in practice, it really reduces flicker to some degree, as you can notice from the above gifs. 

> Maybe an inhibit-redisplay around the deletion insertion call sequence would help?
IIRC I did test adding `(let ((inhibit-redisplay t)) ` around insertation and deletion, there was no difference.  But I guess we can do more tests.

**Comment by minad** [comment]
> I think it's an improvement given that it saves an insertation and a deletion, and the content of this insertation and deletion is longer ( sometimes much longer) than the prefix. 

It might be a slight improvement in terms of code simplification (or maybe not?), but given that the other convention via the `:exit-function` is already well-established, the benefits of this addition should be considered carefully. 

> And in practice, it really reduces flicker to some degree, as you can notice from the above gifs.

But why? Emacs should not randomly initiate redisplay. There must be some bug somewhere, either in your Capf or in Corfu or in Company. My suspicion is that the redisplay is caused by the closing of the popup (either child frame or overlay-based popup) and the exit function is called at some unfortunate point which leads to the flicker. Howerver in Corfu, hiding the popup is even deferred with a timer for performance reasons such that it should always happen afterwards.

1. Selection: add -> addRound...
2. *Unwanted redisplay happens here*
3. :exit-function is called
4. addRound... is deleted
5. Actual candidate is inserted
6. Actually desired redisplay happens here

If the unwanted redisplay could be avoided the observed flicker would go away.

**Comment by zbelial** [comment]
Hi @minad , I agree if we can find why this flicker happens and fix it or prevent it from happening, it will be much better than adding a :company-insert-function ( :company-insert-function may be useful in some other cases, but that's another story). 

So I just made some other tests. It seems that the flicker is related to redisplay.  Here is how I tested. 

Last time I modified `:exit-function` to add `inhibit-redisplay` around the insertation/deletion (Which didn't  inhibit unwanted redisplay in your step 2).  
This time I modified `company-finish` as following:
```elisp
(defun lspce-pre-redisplay (w)
  (message "pre-redisplay lspce"))

(defun company-finish (result)
  (let ((inhibit-redisplay t))
    (message "company-finish...")
    (add-hook 'pre-redisplay-functions #'lspce-pre-redisplay)
    (company--insert-candidate result)
    (company-cancel result)
    (remove-hook 'pre-redisplay-functions #'lspce-pre-redisplay)))
```
Here is a gif. No flicker like before anymore. 

![inhibit-redisplay](https://github.com/company-mode/company-mode/assets/6436456/bf3b6d3a-c241-40d7-a08f-50cad007db8e)

Then I removed inhibit-redisplay from `company-finish`:
```elisp
(defun company-finish (result)
  (message "company-finish...")
  (add-hook 'pre-redisplay-functions #'lspce-pre-redisplay)
  (company--insert-candidate result)
  (company-cancel result)
  (remove-hook 'pre-redisplay-functions #'lspce-pre-redisplay))
```
Flicker happened again. 

![no-inhibit-redisplay](https://github.com/company-mode/company-mode/assets/6436456/a3b14610-c07b-4965-accb-22e77b123f9b)

Redisplay executed several times in each completion. Here is the log:
```elisp
company-finish...
pre-redisplay lspce [8 times]
company-finish...
pre-redisplay lspce [9 times]
```

Any advice on how to figure out why redisply happens?  Thanks.

**Comment by dgutov** [comment]
@zbelial Does your post-completion code contain any process calls or network interactions?

Those can trigger redisplay unless it's inhibited.

**Comment by dgutov** [comment]
@minad As I see it, `company-insert-function` (and corresponding `insert` backend action) could help with two things:

* Conveying to the backend whether the suffix should be replaced as well. Overall the insertion is a less-trivial operation when LSP is involved, and I'm not sure the straight removal of the suffix would work ideally.
* Handling of backends that returns prefixes (and possibly suffixes) of different lengths in `company--multi-backend-adapter`.

**Comment by minad** [comment]
@dgutov You may be right, but unfortunately I don't understand both the scenarios you mention well enough. I think an `:exit-function` can do the equivalent operation as an `:insert-function` in the LSP case as proposed by @zbelial. I am not familiar with additional special behavior of Company. The same with the multi adapter. In `cape-super-capf` (the Cape multi adapter) `:exit-function` is sufficient, but there the prefixes of the merged Capfs must match.

**Comment by dgutov** [comment]
> My post-completion writes to/reads from a lsp server process via an Emacs module

I'm not sure whether that's supposed to trigger redisplay or not. But I guess it might (depends on which primitives the module uses?). So binding `inhibit-redisplay` is one solution, and it's not a terrible one.

**Comment by dgutov** [comment]
> In cape-super-capf (the Cape multi adapter) :exit-function is sufficient, but there the prefixes of the merged Capfs must match.

Perhaps an advanced "merged" exit-function could even deal with that situation.

But I guess that leaves the problem of conveying "insert or replace" intention to the completion table. Right now lsp-mode just has a user option and check the current command, which is decidedly non-optimal.

**Comment by zbelial** [comment]
> So binding `inhibit-redisplay` is one solution, and it's not a terrible one.

Yes, binding `inhibit-redisplay` is a good enough solution for my problem. Thanks.
